### PR TITLE
Add Topology Edge Selection w/ Side Panel

### PR DESCRIPTION
**Pull request by jeff-phillips-18**
Addresses story: https://issues.redhat.com/browse/ODC-2354
Fixes https://issues.redhat.com/browse/ODC-2797
Fixes https://issues.redhat.com/browse/ODC-2726

image

image

image

image

image

image

/cc @openshift/team-devconsole-ux

### CODE REVIEW

**Comment by jeff-phillips-18** [comment]
> @jeff-phillips-18 this looks great, you don't have the Actual traffic connector shown here, but I assume that's coming in a follow up PR?

Correct, we do not have those connectors in yet. see #4026

**Comment by christianvogt** [comment]
Used the `Move` action on a the `connects-to` connector and got many warnings / errors logged in the console. The connection did move in topology though.

The `source` node is showing up in the target dropdown of the `Move` action. This implies i'm going to reconnect the to the source node. This isn't allow. We should remove the source as an option in the menu.

I don't think we should disable the current target node in the `Move` action target dropdown. Simply perform no action if the user keeps the current target selected.

Unable to create a service binding request anymore. The tooltip shows it wants to create an SBR but on release, I get an error in in the console and a visual connector is created instead.

**Comment by jeff-phillips-18** [comment]
Updated as suggested. Not seeing errors any more. Still see some warnings due to https://github.com/patternfly/patternfly-react/issues/3549

**Comment by nemesis09** [comment]
@jeff-phillips-18 
The `move` action on a connector allows event source as a valid dropdown option for a deployment. and throws an error on submit (`Move`)

**Comment by nemesis09** [comment]
The `move` action for a visual connector allows database as valid dropdown option and throws an error on submit (`move`). It updates the connection to database on topology though and changes the connector to type `binding`

**Comment by jeff-phillips-18** [comment]
@nemesis09 Did you try that after my last push?

**Comment by nemesis09** [comment]
> @nemesis09 Did you try that after my last push?

Sorry, I didn't try it after the latest push when I got the errors.
But now that I have pulled in the latest push-
* the form shows an error message on submit for `event source` and `database` as target, but the error message doesn't describe the error and rather shows `error is undefined`
* although the form shows error message for `database` as target on submit, the connector on topology is updated to connect to `database` as `binding connector`
* the `Move` button is disabled after submitting with `database` or `event source`, but it doesn't get enabled again on selecting other valid options from the dropdown. We need to launch it using the `Move` action again.

**Comment by jeff-phillips-18** [comment]
Fixed modal to show the correct error.

Errors updated the edges are reported already in https://issues.redhat.com/browse/ODC-2726 which is fixed in https://github.com/openshift/console/pull/3986

**Comment by nemesis09** [comment]
The modal doesn't show any error now, and submitting(`Move`) with Event Source as a target, just turns out to be a no op

**Comment by nemesis09** [comment]
I tried locally with changes pulled from https://github.com/openshift/console/pull/3986, 
this solves the errors for moving connection from a deployment as visual connector to a database as a binding connector

**Comment by jeff-phillips-18** [comment]
Removed event sources from the list of available targets..

**Comment by nemesis09** [comment]
The modal disables the submit(`move`) button on error, but doesn't enable it back if a different target is selected from the dropdown.

**Comment by nemesis09** [comment]
tested locally. works as expected.
/lgtm

File: frontend/packages/dev-console/src/components/topology/actions/edgeActions.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  return {
+    label: 'Move Connector',
+    callback: () => {
+      moveConnectionModal({ edge, availableTargets });
+    },
+    accessReview: asAccessReview(resourceModel, resourceObj, 'delete'),
+  };
+};
+
+const deleteConnection = (edge: BaseEdge) => {
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/topology/actions/edgeActions.ts]:
```suggestion
const deleteConnection = (edge: BaseEdge): KebabOption => {
```

We need to make sure we type output of newly created objects to insure they stay properly typed (if anything in the type changes later).

File: frontend/packages/dev-console/src/components/topology/actions/edgeActions.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  TYPE_EVENT_SOURCE,
+  TYPE_EVENT_SOURCE_LINK,
+  TYPE_KNATIVE_REVISION,
+  TYPE_KNATIVE_SERVICE,
+  TYPE_REVISION_TRAFFIC,
+  TYPE_SERVICE_BINDING,
+} from '../const';
+import { moveConnectionModal } from '../components/MoveConnectionModal';
+
+const moveConnection = (edge: BaseEdge, availableTargets: Node[]) => {
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/topology/actions/edgeActions.ts]: Same here...
```suggestion
const moveConnection = (edge: BaseEdge, availableTargets: Node[]): KebabOption => {
```

File: frontend/packages/dev-console/src/components/topology/actions/edgeActions.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    actions.push(moveConnection(edge, availableTargets));
+  }
+
+  switch (edge.getType()) {
+    case TYPE_CONNECTS_TO:
+    case TYPE_SERVICE_BINDING:
+      actions.push(deleteConnection(edge));
+      break;
+    default:
+      break;
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/topology/actions/edgeActions.ts]:
```suggestion

```
Default clauses almost never need a break unless you're doing conditional paths within it.

File: frontend/packages/dev-console/src/components/topology/components/MoveConnectionModal.tsx [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+const mapStateToProps = (state: RootState): StateProps => {
+  return {
+    serviceBinding: getServiceBindingStatus(state),
+  };
+};
+
+const ConnectedMoveConnectionModal = connect(mapStateToProps)(MoveConnectionModal);
+
+export const moveConnectionModal = createModalLauncher((props: MoveConnectionModalProps) => (
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/topology/components/MoveConnectionModal.tsx]: 😕 There are a lot of different ways to make modals in our codebase. I'll need to take an action up to talk to Christian and find out what the best way is and try to centralize around it.

**Comment by jeff-phillips-18** [comment]
Added task https://issues.redhat.com/browse/ODC-2823

**Comment by andrewballantyne** [comment]
/retest

Looks like a flake... 
> Kubernetes resource CRUD operations Pod : displays a YAML editor for creating a new resource instance

**Comment by spadgett** [comment]
/hold
for merge queue fix #4065

**Pull request merged**